File: iteration2/src/Transcript.java
package iteration2.src;

import java.util.ArrayList;
import java.util.List;

import iteration2.src.course.*;
import iteration2.src.human.Grade;


public class Transcript {
    private List<CourseRecord> takenCourseRecords;
    //GPA FALAN BASTIRILICAK
    //TODO:add constructor and take course records as parameters as they'll be parsed from json data, and call calculateGPA inside the constructor
    public Transcript(List<CourseRecord> takenCourseRecords){
        this.takenCourseRecords = takenCourseRecords;

        calculateGPA();
    }

    public void addCourseRecord(Course course, LetterGrade lGrade, Season season, Float score, Grade grade, Boolean isPassed){
        
        CourseRecord courserecord = new CourseRecord(course,lGrade,season,grade, score, isPassed);
        takenCourseRecords.add(courserecord);

    }


    public Float calculateGPA(){
        float gpa=0;
        int credits=0;
        float temp=0;
        for(CourseRecord r: takenCourseRecords){
            temp+=r.getlGrade().getNumVal()*r.getCourse().getCredits();
            credits+=r.getCourse().getCredits();
        }
        gpa=temp/credits;
        return gpa;
    }

    public int getCompletedCredits(){
        int completedCredits=0;
        
        for (CourseRecord r: takenCourseRecords){
            if(r.getIsPassed()){
                completedCredits += r.getCourse().getCredits();
            }
        }
        return completedCredits;
    }

    public Boolean checkIfPrerequisitesArePassed(Course course){

        for (Course c : course.getPrerequisites()){
            if(!didStudentPass(course)){
                return false;
            }
        }

        return true;
    }

    public Boolean didStudentPass(Course course){

        Boolean passed = true;

        CourseRecord mostRecent = null;
        for (CourseRecord record : takenCourseRecords){
            if(record.getCourse() == course){
                mostRecent = record;
            }
        }

        if(mostRecent == null)
            return false;

        return mostRecent.getIsPassed();
    }

    public List<CourseRecord> getTakenCourseRecords(){
        return takenCourseRecords;
    }

    public int getNumberOfNTElectivesPassed(){
        int i=0;
        for(CourseRecord r:takenCourseRecords){
            Course course = r.getCourse();
            if(course instanceof NonTechnicalElectiveCourse && didStudentPass(course)){
                i++;
            }
        }

        return i;
    }
    public List<CourseRecord> getNonGradedCourses(){
        List<CourseRecord> courseRecordList = new ArrayList<>();
        for(CourseRecord r:takenCourseRecords){
            if(r.getlGrade() == LetterGrade.NOT_GRADED){
             courseRecordList.add(r);
            }
        }
        return courseRecordList;
    }

    public static LetterGrade getLetterGradeOfScore(float score) {
        if(score >= 0.0f && score < 35.0f){return LetterGrade.FF;}
        else if(score>=35.0f && score < 40.0f ){return  LetterGrade.FD;}
        else if(score>=40.0f && score < 50.0f){return  LetterGrade.DD;}
        else if(score>=50.0f && score < 65.0f){return  LetterGrade.DC;}
        else if(score>=65.0f && score < 75.0f){return  LetterGrade.CC;}
        else if(score>=75.0f && score < 80.0f){return  LetterGrade.CB;}
        else if(score>=80.0f && score < 85.0f){return  LetterGrade.BB;}
        else if(score>=85.0f && score < 90.0f){return  LetterGrade.BA;}
        else{return  LetterGrade.AA;}
    }
    public int getNumberOfTElectivesPassed(){
        int i=0;
        for(CourseRecord r:takenCourseRecords){
            Course course = r.getCourse();
            if(course instanceof TechnicalElectiveCourse && didStudentPass(course)){
                i++;
            }
        }
        return i;
    }
    public int getNumberOfFTElectivesPassed(){
        int i=0;
        for(CourseRecord r:takenCourseRecords){
            Course course = r.getCourse();
            if(course instanceof FacultyTechnicalElectiveCourse && didStudentPass(course)){
                i++;
            }
        }
        return i;
    }
    public boolean didStudentFailBefore(Course course){
        boolean didFail=false;

        for(CourseRecord r:takenCourseRecords) {
            if(r.getCourse()==course)
                didFail=!r.getIsPassed();
        }
        return didFail;
    }
}

File: iteration3/Transcript.py
from iteration3.CourseRecor import CourseRecord
from iteration3.LectureGrade import LetterGrade


class Transcript:
    def __init__(self, taken_course_records):
        self.taken_course_records = taken_course_records
        self.calculate_gpa()
    
    def add_course_record(self, course, l_grade, season, score, grade, is_passed):
        course_record = CourseRecord(course, l_grade, season, grade, score, is_passed)
        self.taken_course_records.append(course_record)
        
    def calculate_gpa(self):
        gpa = 0
        credits = 0
        temp = 0
        for r in self.taken_course_records:
            temp += r.l_grade.num_val * r.course.credits
            credits += r.course.credits
        gpa = temp / credits
        return gpa
    
    def get_completed_credits(self):
        completed_credits = 0
        for r in self.taken_course_records:
            if r.is_passed:
                completed_credits += r.course.credits
        return completed_credits

    def check_if_prerequisites_are_passed(self, course):
        for c in course.prerequisites:
            if not self.did_student_pass(c):
                return False
        return True

    def did_student_pass(self, course):
        passed = True
        most_recent = None
        for record in self.taken_course_records:
            if record.course == course:
                most_recent = record
        if most_recent is None:
            return False
        return most_recent.is_passed

    def get_taken_course_records(self):
        return self.taken_course_records

    def get_non_graded_courses(self):
        course_record_list = []
        for r in self.taken_course_records:
            if r.l_grade == LetterGrade.NOT_GRADED:
                course_record_list.append(r)
        return course_record_list

    @staticmethod
    def get_letter_grade_of_score(score):
        if score >= 0.0 and score < 35.0:
            return LetterGrade.FF
        elif score >= 35.0 and score < 40.0:
            return LetterGrade.FD
        elif score >= 40.0 and score < 50.0:
            return LetterGrade.DD
        elif score >= 50.0 and score < 65.0:
            return LetterGrade.DC
        elif score >= 65.0 and score < 75.0:
            return LetterGrade.CC
        elif score >= 75.0 and score < 80.0:
            return LetterGrade.CB
        elif score >= 80.0 and score < 85.0:
            return LetterGrade.BB
        elif score >= 85.0 and score < 90.0:
            return LetterGrade.BA
        else:
            return LetterGrade.AA


    def get_number_of_te_courses_passed(self):
        i = 0
        for r in self.taken_course_records:
            course = r.course
            if isinstance(course, TechnicalElectiveCourse) and self.did_student_pass(course):
                i += 1
        return i

    def get_number_of_fte_courses_passed(self):
        i = 0
        for r in self.taken_course_records:
            course = r.course
            if isinstance(course, FacultyTechnicalElectiveCourse) and self.did_student_pass(course):
                i += 1
        return i

    def get_number_of_nte_courses_passed(self):
        i = 0
        for r in self.taken_course_records:
            course = r.course
            if isinstance(course, NonTechnicalElectiveCourse) and self.did_student_pass(course):
                i += 1
        return i

    def did_student_fail_before(self, course):
        did_fail = False
        for r in self.taken_course_records:
            if r.course == course:
                did_fail = not r.is_passed
        return did_fail